libevmasm/Inliner.cpp

```diff
@@ -0,0 +1,216 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+/**
+ * @file Inliner.cpp
+ * Inlines small code snippets by replacing JUMP with a copy of the code jumped to.
+ */
+
+#include <libevmasm/Inliner.h>
+
+#include <libevmasm/AssemblyItem.h>
+#include <libevmasm/GasMeter.h>
+#include <libevmasm/KnownState.h>
+#include <libevmasm/SemanticInformation.h>
+
+#include <libsolutil/CommonData.h>
+
+#include <range/v3/numeric/accumulate.hpp>
+#include <range/v3/view/drop_last.hpp>
+#include <range/v3/view/enumerate.hpp>
+#include <range/v3/view/slice.hpp>
+#include <range/v3/view/transform.hpp>
+
+#include <optional>
+
+
+using namespace std;
+using namespace solidity;
+using namespace solidity::evmasm;
+
+bool Inliner::isInlineCandidate(u256 const& _tag, InlinableBlock const& _block) const
+{
+	assertThrow(_block.items.size() > 0, OptimizerException, "");
+
+	// Never inline tags that reference themselves.
```

Comment:
No, bad it'll blow up code...
```
tag_1:
  CALLVALUE
  PUSHTAG(tag_1)
  JUMP
```

after one excution:

```
tag_1:
  CALLVALUE
  CALLVALUE
  PUSHTAG(tag_1)
  JUMP
```

and so on exponentially...

But: since we restrict to only jump-in jump-out this will never happen anyways. And even if we were to enable more combinations, it'd always be infinite-loop situations... so not that dramatic...